ovs/lib/mdsservice.py

```diff
@@ -156,92 +160,107 @@ def prepare_mds_service(storagerouter, vpool, fresh_only):
-            storagedriver_config.configure_metadata_server(mds_nodes=mds_nodes)
-            storagedriver_config.save(client)
-        finally:
-            prepare_mds_lock.release()
-
+            MDSServiceController._logger.info('Prepare MDS: StorageRouter {0} - vPool {1}: Configuring StorageDriver with MDS nodes: {2}'.format(storagerouter.name, vpool.name, mds_nodes))
+            # Generate the correct section in the StorageDriver's configuration
+            try:
+                storagedriver_config = StorageDriverConfiguration('storagedriver', vpool.guid, storagedriver.storagedriver_id)
+                storagedriver_config.load()
+                storagedriver_config.configure_metadata_server(mds_nodes=mds_nodes)
+                storagedriver_config.save(client)
+            except Exception:
+                MDSServiceController._logger.exception('Prepare MDS: StorageRouter {0} - vPool {1}: Configuring StorageDriver failed. Reverting model changes'.format(storagerouter.name, vpool.name))
+                # Clean up model changes if error occurs
+                sdp.delete()
+                mds_service.delete()  # Must be removed before the service
+                service.delete()
         return mds_service
 
     @staticmethod
-    def remove_mds_service(mds_service, vpool, reconfigure, allow_offline=False):
+    def remove_mds_service(mds_service, reconfigure, allow_offline=False):
         """
         Removes an MDS service
         :param mds_service: The MDS service to remove
         :type mds_service: ovs.dal.hybrids.j_mdsservice.MDSService
-        :param vpool: The vPool for which the MDS service will be removed
-        :type vpool: ovs.dal.hybrids.vpool.VPool
         :param reconfigure: Indicates whether reconfiguration is required
         :type reconfigure: bool
         :param allow_offline: Indicates whether it's OK that the node for which mds services are cleaned is offline
         :type allow_offline: bool
+        :raises: RuntimeError - When vDisks present on the MDSService to be removed
+                              - No StorageDriver is linked to the MDSService to be removed
+                 UnableToConnectException - When StorageRouter on which the MDSService resides is unreachable and allow_offline flag is False
+        :return: None
+        :rtype: NoneType
         """
-        if mds_service.guid not in vpool.mds_services_guids:
-            raise RuntimeError('MDS Service {0} is not related to vPool {1}'.format(mds_service.service.name, vpool.name))
-
         if len(mds_service.vdisks_guids) > 0 and allow_offline is False:
             raise RuntimeError('Cannot remove MDSService that is still serving disks')
 
         if len(mds_service.storagedriver_partitions) == 0 or mds_service.storagedriver_partitions[0].storagedriver is None:
             raise RuntimeError('Failed to retrieve the linked StorageDriver to this MDS Service {0}'.format(mds_service.service.name))
```

Comment:
On Nuvolat we had the case where the Storagedriver was removed and the mds_service remained in the model.
This way, the service could not get cleanup up in conventional ways.
Not sure if we should tackle with removing our check here though.

Reply:
Removing a StorageDriver should first check whether above conditions are met before actually doing anything. These checks should however remain in remove_mds_service imo